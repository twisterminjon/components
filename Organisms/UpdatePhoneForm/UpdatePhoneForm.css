.updatephoneform {
  padding: 16px;
  background-color: var(--menu__bg);
}

.updatephoneform h1 {
  font-size: var(--font-base-headline--size);
  font-weight: var(--font-base--bold);
}

.updatephoneform-input {
  padding: 0 0 160px 0;
}

.updatephoneform-loader {
  width: 100%;
  padding: 80px 0 80px 0;
}

.ui.modal .actions > .button.updatephoneform-button {
  width: 100%;
  margin: 0 0 10px 0;
}
