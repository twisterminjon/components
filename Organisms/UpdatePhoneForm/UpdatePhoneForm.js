import React, { useState } from 'react'
import PropTypes from 'prop-types'

import { Modal } from 'semantic-ui-react'

import Button from '../../Atoms/Button/Button'
import ButtonGhost from '../../Atoms/ButtonGhost/ButtonGhost'
import SpinnerDots from '../../Atoms/SpinnerDots/SpinnerDots'
import PhoneInput from '../../Molecules/PhoneInput/PhoneInput'
import { isPhoneNumberValid, parsePhoneNumber } from '@shared/helpers'
import ErrorMessages from '../../../ErrorMessages.json'
import ErrorPage from '../../Views/ErrorPage/ErrorPage'

import './UpdatePhoneForm.css'

UpdatePhoneForm.propTypes = {
  /** If true the form will be displayed */
  show: PropTypes.bool.isRequired,

  /** Current user id */
  currentUserId: PropTypes.string,

  /** Current phone number of the user */
  currentPhone: PropTypes.string,

  /** Can show a loader when loading phone data */
  phoneLoading: PropTypes.bool,

  /** Can show a loader when running mutation */
  updateLoading: PropTypes.bool,

  /** Called after phone is update*/
  onUpdatePhone: PropTypes.func.isRequired,

  /** Called when the close button is clicked */
  onClose: PropTypes.func.isRequired,
}

export default function UpdatePhoneForm({
  show,
  currentUserId,
  phoneLoading,
  updateLoading,
  onUpdatePhone,
  currentPhone = '',
  onClose,
}) {
  const [phoneValue, setPhoneValue] = useState(parsePhoneNumber(currentPhone, 'US').international)
  const [error, setError] = useState('')

  const handleUpdatePhone = () => {
    if (isPhoneNumberValid(phoneValue)) {
      setError('')
      onUpdatePhone({ variables: { id: currentUserId, phone: phoneValue } })
        .then(() => handleCloseModal())
        .catch(e => {
          if (e.message.includes(ErrorMessages.Create.INVALID_PHONE_FORMAT)) {
            setError("Phone number doesn't exist or is invalid")
          } else {
            return <ErrorPage error={e} />
          }
        })
    } else {
      setError("Phone number doesn't exist or is invalid")
    }
  }

  const handleCloseModal = () => {
    setPhoneValue(parsePhoneNumber(currentPhone, 'US').international)
    setError('')
    onClose()
  }

  return (
    <Modal open={show} onClose={handleCloseModal} size="tiny" basic centered={false}>
      <div className="updatephoneform" data-testid="updatephoneform">
        <h1 data-testid="updatephoneform-title">Update Phone Number</h1>
        <Modal.Content>
          <Modal.Description>
            {phoneLoading ? (
              <SpinnerDots className="updatephoneform-loader" />
            ) : (
              <div className="updatephoneform-input">
                <PhoneInput
                  label="Phone Number"
                  name="phone"
                  value={phoneValue}
                  onChange={setPhoneValue}
                  hasError={!!error}
                  errorMessage={error}
                  key="update-phone-input"
                  data-testid="updatephoneform-input"
                />
              </div>
            )}
          </Modal.Description>
          <Modal.Actions>
            <Button
              data-testid="updatephoneform-button-continue"
              disabled={phoneValue === currentPhone || !phoneValue.trim() || updateLoading}
              className="updatephoneform-button"
              size="big"
              fluid
              loading={updateLoading}
              onClick={() => handleUpdatePhone()}>
              Continue
            </Button>
            <ButtonGhost
              data-testid="updatephoneform-button-cancel"
              className="updatephoneform-button"
              size="big"
              fluid
              onClick={handleCloseModal}>
              Cancel
            </ButtonGhost>
          </Modal.Actions>
        </Modal.Content>
      </div>
    </Modal>
  )
}
