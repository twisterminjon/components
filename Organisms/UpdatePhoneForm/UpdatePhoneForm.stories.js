import React from 'react'

import { storiesOf } from '@storybook/react'
import StoryRouter from 'storybook-react-router'
import { withKnobs, boolean, text } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import UpdatePhoneForm from './UpdatePhoneForm'

const stories = storiesOf('Provider/v1/Organisms', module)

stories
  .addDecorator(withKnobs)
  .addDecorator(StoryRouter())
  .add('UpdatePhoneForm', () => {
    const showVal = boolean('show', true)
    const mutationLoadingVal = boolean('mutationLoading', false)
    const phoneLoadingVal = boolean('phoneLoading', false)
    const phoneVal = text('phone', '+17272622331')

    return (
      <UpdatePhoneForm
        show={showVal}
        onClose={action('Close')}
        currentPhone={phoneVal}
        updateLoading={mutationLoadingVal}
        phoneLoading={phoneLoadingVal}
        onUpdatePhone={action('Update Phone')}
        updateError={''}
      />
    )
  })
