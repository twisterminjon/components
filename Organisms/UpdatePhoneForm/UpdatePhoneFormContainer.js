import React, { useContext, useEffect } from 'react'
import { NetworkStatus } from 'apollo-client'
import PropTypes from 'prop-types'

import { useMutation, useQuery } from 'react-apollo'
import UsersQl from '../../../services/UsersQl'
import { CurrentUserContext } from '@shared/providers'

import ErrorPage from '../../Views/ErrorPage/ErrorPage'
import UpdatePhoneForm from './UpdatePhoneForm'

UpdatePhoneFormContainer.propTypes = {
  /** Can show or be hidden */
  show: PropTypes.bool,

  /** Called when the close button is clicked */
  onClose: PropTypes.func.isRequired,
}

UpdatePhoneFormContainer.defaultProps = {
  show: false,
}

const GET_USER_INFO = UsersQl.getByIdShort()
const PHONE_MUTATION = UsersQl.updatePhone()

export default function UpdatePhoneFormContainer({ show, onClose }) {
  const currentUser = useContext(CurrentUserContext)
  const { data, error, refetch, networkStatus } = useQuery(GET_USER_INFO, {
    fetchPolicy: 'network-only',
    variables: { id: currentUser.id },
    skip: currentUser.id < 0,
    notifyOnNetworkStatusChange: true,
  })

  useEffect(() => {
    refetch()
  }, [show, refetch])

  const [updatePhoneNumber, { loading: mutationLoading }] = useMutation(PHONE_MUTATION)

  if (error) {
    return <ErrorPage error={error} />
  }

  const phone = data ? data.user.phone : ''

  return (
    <UpdatePhoneForm
      key={phone}
      show={show}
      currentUserId={currentUser.id}
      onClose={onClose}
      currentPhone={phone}
      phoneLoading={NetworkStatus.refetch === networkStatus}
      updateLoading={mutationLoading}
      onUpdatePhone={updatePhoneNumber}
    />
  )
}
