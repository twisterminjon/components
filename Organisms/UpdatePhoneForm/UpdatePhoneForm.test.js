import React from 'react'
import UpdatePhoneForm from './UpdatePhoneForm'

describe('UpdatePhoneForm', () => {
  const mockFun = jest.fn()

  it('matches the snapshot', () => {
    const wrapper = window.shallow(
      <UpdatePhoneForm
        show={true}
        onClose={mockFun}
        currentPhone={'+17272622331'}
        updateLoading={false}
        onUpdatePhone={mockFun}
        updateError={''}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('matches the snapshot when hidden', () => {
    const wrapper = window.shallow(
      <UpdatePhoneForm
        show={false}
        onClose={mockFun}
        currentPhone={'+17272622331'}
        updateLoading={false}
        onUpdatePhone={mockFun}
        updateError={''}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('matches the snapshot when saving phone', () => {
    const wrapper = window.shallow(
      <UpdatePhoneForm
        show={true}
        onClose={mockFun}
        currentPhone={'+17272622331'}
        updateLoading={true}
        onUpdatePhone={mockFun}
        updateError={''}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('matches the snapshot when getting phone', () => {
    const wrapper = window.shallow(
      <UpdatePhoneForm
        show={true}
        onClose={mockFun}
        currentPhone={'+17272622331'}
        updateLoading={false}
        phoneLoading={true}
        onUpdatePhone={mockFun}
        updateError={''}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
