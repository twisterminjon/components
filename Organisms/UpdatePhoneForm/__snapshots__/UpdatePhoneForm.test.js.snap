// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePhoneForm matches the snapshot 1`] = `
<Modal
  basic={true}
  centered={false}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={true}
  size="tiny"
>
  <div
    className="updatephoneform"
    data-testid="updatephoneform"
  >
    <h1
      data-testid="updatephoneform-title"
    >
      Update Phone Number
    </h1>
    <ModalContent>
      <ModalDescription>
        <div
          className="updatephoneform-input"
        >
          <PhoneInput
            alertMessage=""
            data-testid="updatephoneform-input"
            disabled={false}
            displayOnly={false}
            errorMessage=""
            hasError={false}
            hintMessage=""
            hintText=""
            key="update-phone-input"
            label="Phone Number"
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyPress={[Function]}
            onValueInit={[Function]}
            placeholder=""
            required={false}
            value="+17272622331"
          />
        </div>
      </ModalDescription>
      <ModalActions>
        <Button
          className="updatephoneform-button"
          data-testid="updatephoneform-button-continue"
          disabled={true}
          fluid={true}
          loading={false}
          onClick={[Function]}
          size="big"
          style={Object {}}
          type="submit"
        >
          Continue
        </Button>
        <ButtonGhost
          className="updatephoneform-button"
          data-testid="updatephoneform-button-cancel"
          fluid={true}
          onClick={[Function]}
          red={false}
          size="big"
          style={Object {}}
        >
          Cancel
        </ButtonGhost>
      </ModalActions>
    </ModalContent>
  </div>
</Modal>
`;

exports[`UpdatePhoneForm matches the snapshot when getting phone 1`] = `
<Modal
  basic={true}
  centered={false}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={true}
  size="tiny"
>
  <div
    className="updatephoneform"
    data-testid="updatephoneform"
  >
    <h1
      data-testid="updatephoneform-title"
    >
      Update Phone Number
    </h1>
    <ModalContent>
      <ModalDescription>
        <SpinnerDots
          className="updatephoneform-loader"
          style={Object {}}
        />
      </ModalDescription>
      <ModalActions>
        <Button
          className="updatephoneform-button"
          data-testid="updatephoneform-button-continue"
          disabled={true}
          fluid={true}
          loading={false}
          onClick={[Function]}
          size="big"
          style={Object {}}
          type="submit"
        >
          Continue
        </Button>
        <ButtonGhost
          className="updatephoneform-button"
          data-testid="updatephoneform-button-cancel"
          fluid={true}
          onClick={[Function]}
          red={false}
          size="big"
          style={Object {}}
        >
          Cancel
        </ButtonGhost>
      </ModalActions>
    </ModalContent>
  </div>
</Modal>
`;

exports[`UpdatePhoneForm matches the snapshot when hidden 1`] = `
<Modal
  basic={true}
  centered={false}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={false}
  size="tiny"
>
  <div
    className="updatephoneform"
    data-testid="updatephoneform"
  >
    <h1
      data-testid="updatephoneform-title"
    >
      Update Phone Number
    </h1>
    <ModalContent>
      <ModalDescription>
        <div
          className="updatephoneform-input"
        >
          <PhoneInput
            alertMessage=""
            data-testid="updatephoneform-input"
            disabled={false}
            displayOnly={false}
            errorMessage=""
            hasError={false}
            hintMessage=""
            hintText=""
            key="update-phone-input"
            label="Phone Number"
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyPress={[Function]}
            onValueInit={[Function]}
            placeholder=""
            required={false}
            value="+17272622331"
          />
        </div>
      </ModalDescription>
      <ModalActions>
        <Button
          className="updatephoneform-button"
          data-testid="updatephoneform-button-continue"
          disabled={true}
          fluid={true}
          loading={false}
          onClick={[Function]}
          size="big"
          style={Object {}}
          type="submit"
        >
          Continue
        </Button>
        <ButtonGhost
          className="updatephoneform-button"
          data-testid="updatephoneform-button-cancel"
          fluid={true}
          onClick={[Function]}
          red={false}
          size="big"
          style={Object {}}
        >
          Cancel
        </ButtonGhost>
      </ModalActions>
    </ModalContent>
  </div>
</Modal>
`;

exports[`UpdatePhoneForm matches the snapshot when saving phone 1`] = `
<Modal
  basic={true}
  centered={false}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={true}
  size="tiny"
>
  <div
    className="updatephoneform"
    data-testid="updatephoneform"
  >
    <h1
      data-testid="updatephoneform-title"
    >
      Update Phone Number
    </h1>
    <ModalContent>
      <ModalDescription>
        <div
          className="updatephoneform-input"
        >
          <PhoneInput
            alertMessage=""
            data-testid="updatephoneform-input"
            disabled={false}
            displayOnly={false}
            errorMessage=""
            hasError={false}
            hintMessage=""
            hintText=""
            key="update-phone-input"
            label="Phone Number"
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyPress={[Function]}
            onValueInit={[Function]}
            placeholder=""
            required={false}
            value="+17272622331"
          />
        </div>
      </ModalDescription>
      <ModalActions>
        <Button
          className="updatephoneform-button"
          data-testid="updatephoneform-button-continue"
          disabled={true}
          fluid={true}
          loading={true}
          onClick={[Function]}
          size="big"
          style={Object {}}
          type="submit"
        >
          Continue
        </Button>
        <ButtonGhost
          className="updatephoneform-button"
          data-testid="updatephoneform-button-cancel"
          fluid={true}
          onClick={[Function]}
          red={false}
          size="big"
          style={Object {}}
        >
          Cancel
        </ButtonGhost>
      </ModalActions>
    </ModalContent>
  </div>
</Modal>
`;
